### PR TITLE
(PDB-1851) Create new "resource-centric" wire-format for reports

### PR DESCRIPTION
**Pull request by ajroetker**
Previously PuppetDB report's wire-format used a flattened
`resource_events`, where resources and events where merged together in
the terminus and stored under the same model in PuppetDB.
This commit adds a new report wire-format which uses a list of
`resources` which each have a field `events` which is a list of events
associated with the resource.

### CODE REVIEW

File: puppet/lib/puppet/reports/puppetdb.rb [hunk 1]
```diff
@@ -48,31 +48,41 @@ def report_to_hash
...
-          events.concat([fabricate_event(status, "skipped")])
-        end
-        events
-      end)
+  def build_resources_list
+    profile("Build resources list (count: #{resource_statuses.count})",
+            [:puppetdb, :resources_list, :build]) do
+      include_unchanged_resources = false
+      resources = resource_statuses.values.map { |resource| resource_status_to_hash(resource) }
+      if ! include_unchanged_resources
```
Comment (wkalt) [inline puppet/lib/puppet/reports/puppetdb.rb]: this is hardcoded false right above -- is that a bug?

Comment (ajroetker) [inline puppet/lib/puppet/reports/puppetdb.rb]: Nope it's another ticket to make it a configurable value, I just named it for now to make that ticket easier when the time comes!

Comment (wkalt) [inline puppet/lib/puppet/reports/puppetdb.rb]: got it

File: puppet/lib/puppet/util/puppetdb/config.rb [hunk 2]
```diff
@@ -56,6 +54,7 @@ def self.load(config_file = nil)
       config_hash = defaults.merge(main_section).reject do |k, v|
         !([:server,
            :port,
+           # TODO :ignore_blacklisted_events needs to be deprecated
```
Comment (wkalt) [inline puppet/lib/puppet/util/puppetdb/config.rb]: was it ever actually user-exposed? my understanding is no

**Comment by wkalt** [comment]
this needs documentation updates

**Comment by wkalt** [comment]
this breaks with this manifest, which submits a skipped event:

```
schedule {'everyday':
  period => daily,
  range => "2-4",
}

notify {"hello":
  message => "hello planet",
  schedule => 'everyday',
}
```

we should be testing that behavior once it's fixed

File: test/puppetlabs/puppetdb/command_test.clj [hunk 3]
```diff
@@ -742,7 +742,7 @@
       (let [result (query-to-vec "SELECT certname,environment_id FROM factsets")]
         (is (= result [(with-env {:certname certname})]))))))
 
-(deftest replace-facts-with-v2-wire-format
+#_(deftest replace-facts-with-v2-wire-format
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/command_test.clj]: left over

**Comment by senior** [comment]
The new report version needs doc changes

**Comment by ajroetker** [comment]
Only docs changes since the last acceptance tests run passed.

**Comment by ajroetker** [comment]
transients ^

File: documentation/api/command/v1/commands.markdown [hunk 4]
```diff
@@ -98,15 +99,32 @@ payload of this command.
 See [deactivate node wire format v3][deactivatev3] for more information on the
 payload of this command.
 
+### "store report", version 6
+
+The version 6 store report command differs from previous versions by changing
+from a `resource_events` property to a `resources` property. The
+`resource_events` property was a merged version of `resources` and their
+associated events `events`. This new version moves the command to use a similar
+format to a raw Puppet report, with a list of `resources` each with a `events`
```
Comment (wkalt) [inline documentation/api/command/v1/commands.markdown]: an events property

File: documentation/api/command/v1/commands.markdown [hunk 5]
```diff
@@ -98,15 +99,32 @@ payload of this command.
...
+[report wire format v6][reportv6].
+
+## Deprecated Commands
+
 ### "store report", version 5
 
-The version 5 store report command differs from version four in the addition of
-a "noop" flag, which is a boolean indicating whether the report was produced by
-a puppet run with --noop, as well as in the conversion of dash-separated fields to
+> **Note:** This version is deprecated, use the latest version instead.
```
Comment (wkalt) [inline documentation/api/command/v1/commands.markdown]: I'd drop this -- it's already under a deprecated header and we don't want to repeat the note for each deprecation.

File: documentation/api/wire_format/report_format_v6.markdown [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        "producer_timestamp": <datetime>,
+        "resources": [<resource>, <resource>, ...],
+        "metrics": [<metric>, <metric>, ...],
+        "logs": [<log>, <log>, ...],
+        "transaction_uuid": <string>,
+        "status": <string>,
+        "noop": <boolean>
+    }
+
+**NOTE** `metrics` and `logs` may also take the value null.
```
Comment (wkalt) [inline documentation/api/wire_format/report_format_v6.markdown]: we can drop this note. it's explained below.

File: documentation/api/wire_format/report_format_v6.markdown [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+      "resource_type": <type of resource>,
+      "resource_title": <title of resource>,
+      "skipped" : <boolean for whether or not the resource was skipped>, 
+      "events" : [<event>, <event>, ...]
+      "file": <manifest file containing resource definition>,
+      "line": <line in manifest file on which resource is defined>,
+      "containment_path": <containment heirarchy of resource within catalog>
+    }
+
+In each `<resource>` object the `"events"` is an array of objects of the following form:
```
Comment (wkalt) [inline documentation/api/wire_format/report_format_v6.markdown]: "the events are" or "events is"

File: documentation/api/wire_format/report_format_v6.markdown [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    {
+     "property": <string>,
+     "timestamp": <datetime>,
+     "status": <string>,
+     "old_value": <string>,
+     "new_value": <string>,
+     "message": <string>,
+     "file": <string>,
+     "line: <integer>
+    }
```
Comment (wkalt) [inline documentation/api/wire_format/report_format_v6.markdown]: this same object is on line 151. We can probably delete it up there and leave it here.

File: puppet/spec/unit/util/puppetdb/config_spec.rb [hunk 9]
```diff
@@ -185,34 +181,34 @@ def write_config(content)
       it "should fail if given an invalid hostname" do
         write_config <<CONF
 [main]
-server_urls = https://invalid_host_name.com
+server_urls = https://invalid_host\[\]_name.com
```
Comment (wkalt) [inline puppet/spec/unit/util/puppetdb/config_spec.rb]: why this change?

File: test/puppetlabs/puppetdb/command_test.clj [hunk 10]
```diff
@@ -1200,6 +1200,34 @@
...
+  [resource-events]
+  (for [[resource events] (group-by resource-event->resource resource-events)]
+    (assoc resource :events (map #(dissoc % :file :line :resource_type :resource_title :containment_path) events))))
+
+(let [report (-> (:basic report-examples/reports)
+                 (assoc :environment "DEV")
+                 munge-example-report-for-storage
+                 (update :resource_events resource-events->resources)
+                 (clojure.set/rename-keys {:resource_events :resources}))
+      _ (println report)
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/command_test.clj]: left over?

**Comment by wkalt** [comment]
@ajroetker a few more minor comments. Assuming we're good on the blacklist stuff I think it looks good otherwise.

File: test/puppetlabs/puppetdb/command_test.clj (hunk 10 above)
Comment (ajroetker) [inline test/puppetlabs/puppetdb/command_test.clj]: Ooops nice catch

File: puppet/spec/unit/util/puppetdb/config_spec.rb (hunk 9 above)
Comment (ajroetker) [inline puppet/spec/unit/util/puppetdb/config_spec.rb]: Ruby 2's URI library doesn't think `invalid_host_name` is invalid, so my tests were failing.

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by wkalt** [comment]
+1
